wrappers/services.go

```diff
@@ -375,120 +376,202 @@ func userDaemonReload() error {
 		}
 
-		// Generate systemd .socket files if needed
-		var socketFiles *map[string][]byte
-		socketFiles, err = generateSnapSocketFiles(app)
-		if err != nil {
-			return err
+		// TODO: do we still need this?
+		if opts.Preseeding {
+			preseeding = true
 		}
-		for path, content := range *socketFiles {
-			os.MkdirAll(filepath.Dir(path), 0755)
-			if err = osutil.AtomicWriteFile(path, content, 0644, 0); err != nil {
-				return err
+
+		for _, app := range s.Apps {
+			if !app.IsService() {
+				continue
 			}
-			written = append(written, path)
-		}
 
-		if app.Timer != nil {
+			handleFile := func(path string, desiredContent []byte) error {
+				// get the existing file state (if any) of the service to have
+				// something to rollback to if we have any errors
+
+				// note we can't use FileReference here since we may be modifying
+				// the file, and the FileReference.State() wouldn't be evaluated
+				// until _after_ we attempted modification
+				oldFileState := osutil.MemoryFileState{}
+				haveOldFile := false
+				st, stErr := os.Stat(path)
+				if stErr == nil {
+					haveOldFile = true
+					b, err := ioutil.ReadFile(path)
+					if err != nil {
+						return err
+					}
+					oldFileState.Content = b
+					oldFileState.Mode = st.Mode()
+				}
+
+				newFileState := osutil.MemoryFileState{
+					Content: desiredContent,
+					Mode:    os.FileMode(0644),
+				}
+				if haveOldFile {
+					// use whatever the old mode was if we have it
+					newFileState.Mode = st.Mode()
+				}
+
+				if mkdirErr := os.MkdirAll(filepath.Dir(path), 0755); mkdirErr != nil {
+					return mkdirErr
+				}
+				ensureErr := osutil.EnsureFileState(path, &newFileState)
+				if ensureErr != nil && ensureErr != osutil.ErrSameState {
+					// fatal error trying to write the file
+					return ensureErr
+				}
+				if ensureErr == nil {
+					// then we modified the file and should save this as something
+					// to roll back to
+					if haveOldFile {
+						st := osutil.FileState(&oldFileState)
+						modifiedUnits[path] = &st
+					} else {
+						modifiedUnits[path] = nil
+					}
+					modified[s] = append(modified[s], app)
+				}
+
+				switch app.DaemonScope {
+				case snap.SystemDaemon:
+					modifiedSystem = true
+				case snap.UserDaemon:
+					modifiedUser = true
+				}
+
+				return nil
+			}
+
+			// create services first; this doesn't trigger systemd
+
+			// Generate new service file state
+			path := app.ServiceFile()
 			var content []byte
-			content, err = generateSnapTimerFile(app)
+			content, err = generateSnapServiceFile(app, opts)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			path := app.Timer.File()
-			os.MkdirAll(filepath.Dir(path), 0755)
-			if err = osutil.AtomicWriteFile(path, content, 0644, 0); err != nil {
-				return err
+
+			err = handleFile(path, content)
+			if err != nil {
+				return nil, err
+			}
+
+			// Generate systemd .socket files if needed
+			var socketFiles *map[string][]byte
```

Comment:
This feels a bit unusual, why a pointer to a map? It seems if I change this code to just use `map[string][]byte` things are fine. What am I missing :) ?

Reply:
yes, it seems to be some historical weirdness, we should fix it

Reply:
this was the previous version of the code, but I can change that too, there doesn't seem to be any real reason to this

Reply:
this was removed